File: cloud/cloud.go
// Package cloud is the resource module for cloud-side provisioning.
//
// These endpoints provision real resources in AWS/GCP/Azure, NOT on the
// SSH-into-VM path that install/ and deploy/ use. Examples:
//
//   - S3 buckets, GCS buckets
//   - IAM roles and policies
//   - VPCs and subnets
//   - EC2 / Compute Engine instances
//   - RDS / Cloud SQL instances
//   - Lambda / Cloud Functions
//   - EKS / GKE clusters
//
// Backed by /api/v2/tenant/provision/* on the tenant node, which in turn
// runs Terraform server-side.
package cloud

import (
	"context"
	"fmt"

	vxerrors "github.com/prodxcloud/vxcloud/errors"
	"github.com/prodxcloud/vxcloud/transport"
)

// Result is what every cloud provisioning endpoint returns.
type Result struct {
	SessionID        string                 `json:"session_id"`
	Status           string                 `json:"status,omitempty"`
	ResourceName     string                 `json:"resource_name,omitempty"`
	ARN              string                 `json:"arn,omitempty"`
	BucketName       string                 `json:"bucket_name,omitempty"`
	InstanceID       string                 `json:"instance_id,omitempty"`
	ExecutionTime    string                 `json:"execution_time,omitempty"`
	StatePath        string                 `json:"state_path,omitempty"`
	TerraformOutputs map[string]interface{} `json:"terraform_outputs,omitempty"`
}

// Client is the cloud facade.
type Client struct {
	T              *transport.Transport
	NodeURL        string
	AuthedUsername string
}

// S3 returns the S3 / object-storage sub-client.
func (c *Client) S3() *S3 { return &S3{c: c} }

// IAM returns the IAM sub-client (roles + policies).
func (c *Client) IAM() *IAM { return &IAM{c: c} }

// VM returns the EC2 / Compute-Engine sub-client.
func (c *Client) VM() *VM { return &VM{c: c} }

// Network returns the VPC / network sub-client.
func (c *Client) Network() *Network { return &Network{c: c} }

// Database returns the RDS / Cloud SQL sub-client.
func (c *Client) Database() *Database { return &Database{c: c} }

// Kubernetes returns the EKS / GKE sub-client.
func (c *Client) Kubernetes() *Kubernetes { return &Kubernetes{c: c} }

// Serverless returns the Lambda / Cloud Functions sub-client.
func (c *Client) Serverless() *Serverless { return &Serverless{c: c} }

// commonProvision sends a provisioning POST to the given path with the
// canonical envelope all /api/v2/tenant/provision/* handlers expect.
func (c *Client) commonProvision(ctx context.Context, op, path, app, cloud, region, env, resourceType string, extras map[string]interface{}) (*Result, error) {
	if env == "" {
		env = "development"
	}
	if cloud == "" {
		cloud = "aws"
	}
	if region == "" {
		region = "us-east-1"
	}
	body := map[string]interface{}{
		"app_name":       app,
		"resource_name":  app,
		"instance_name":  app,
		"network_name":   app,
		"key_name":       app,
		"role_name":      app,
		"hostname":       app,
		"cloud_provider": cloud,
		"region":         region,
		"environment":    env,
		"resource_type":  resourceType,
		"username":       c.AuthedUsername,
	}
	for k, v := range extras {
		body[k] = v
	}
	u := transport.JoinURL(c.NodeURL, path)
	var out Result
	if err := c.T.JSON(ctx, op, "POST", u, body, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// ── S3 / Storage ──────────────────────────────────────────────────

type S3 struct{ c *Client }

// CreateBucket provisions a bucket. Cloud may be "aws" (S3) or "gcp" (GCS)
// or "azure" (Blob); region is the cloud-specific region string.
func (s *S3) CreateBucket(ctx context.Context, name, cloud, region string) (*Result, error) {
	return s.c.commonProvision(ctx,
		"cloud.S3.CreateBucket",
		"/api/v2/tenant/provision/storage",
		name, cloud, region, "", "s3", nil)
}

// ── IAM ────────────────────────────────────────────────────────────

type IAM struct{ c *Client }

// CreatePolicy provisions an IAM policy. policyDocument is the inline JSON
// document; the SDK passes it through to the Terraform aws_iam_policy
// resource.
func (i *IAM) CreatePolicy(ctx context.Context, name string, policyDocument string) (*Result, error) {
	return i.c.commonProvision(ctx,
		"cloud.IAM.CreatePolicy",
		"/api/v2/tenant/provision/security",
		name, "aws", "", "", "policy",
		map[string]interface{}{"policy_document": policyDocument},
	)
}

// CreateRole provisions an IAM role. assumeRolePolicy is the trust policy
// document.
func (i *IAM) CreateRole(ctx context.Context, name string, assumeRolePolicy string) (*Result, error) {
	return i.c.commonProvision(ctx,
		"cloud.IAM.CreateRole",
		"/api/v2/tenant/provision/security",
		name, "aws", "", "", "role",
		map[string]interface{}{"assume_role_policy": assumeRolePolicy},
	)
}

// CreateKeypair provisions an EC2 key pair. The private key is stored in
// the workspace vault under the given name; the public material is
// uploaded to AWS.
func (i *IAM) CreateKeypair(ctx context.Context, name, region string) (*Result, error) {
	return i.c.commonProvision(ctx,
		"cloud.IAM.CreateKeypair",
		"/api/v2/tenant/provision/security",
		name, "aws", region, "", "keypair", nil)
}

// ── VM (EC2 / Compute Engine) ─────────────────────────────────────

type VM struct{ c *Client }

// VMOpts describes a VM provision request.
type VMOpts struct {
	InstanceType string // e.g. t3.small
	VolumeSizeGB int
	Replicas     int
}

// Create provisions a VM.
func (v *VM) Create(ctx context.Context, name, cloud, region string, opts VMOpts) (*Result, error) {
	if opts.InstanceType == "" {
		opts.InstanceType = "t2.micro"
	}
	if opts.VolumeSizeGB == 0 {
		opts.VolumeSizeGB = 30
	}
	if opts.Replicas == 0 {
		opts.Replicas = 1
	}
	return v.c.commonProvision(ctx,
		"cloud.VM.Create",
		"/api/v2/tenant/provision/vm",
		name, cloud, region, "", "vm",
		map[string]interface{}{
			"instance_type": opts.InstanceType,
			"volume_size":   opts.VolumeSizeGB,
			"replicas":      opts.Replicas,
		},
	)
}

// StatusInput describes a VM status lookup.
type StatusInput struct {
	// InstanceID is the cloud provider's instance identifier (e.g. an
	// AWS EC2 instance id). Required.
	InstanceID string
	// Cloud is the provider ("aws" | "gcp" | "azure"). Defaults to "aws".
	Cloud string
}

// ActionInput describes a VM lifecycle action.
type ActionInput struct {
	// InstanceID is the cloud provider's instance identifier. Required.
	InstanceID string
	// Action is one of "start", "stop", "restart", "reboot". Required.
	Action string
	// Cloud is the provider ("aws" | "gcp" | "azure"). Defaults to "aws".
	Cloud string
}

// Status returns the current power / lifecycle state of a previously-
// provisioned VM.
//
// Maps to POST {NodeURL}/api/v2/provision/vm/status.
func (v *VM) Status(ctx context.Context, input StatusInput) (map[string]any, error) {
	if input.InstanceID == "" {
		return nil, &vxerrors.ValidationError{Failure: &vxerrors.Failure{
			Op: "cloud.VM.Status", Message: "InstanceID is required",
		}}
	}
	provider := input.Cloud
	if provider == "" {
		provider = "aws"
	}
	url := transport.JoinURL(v.c.NodeURL, "/api/v2/provision/vm/status")
	body := map[string]any{
		"instance_id": input.InstanceID,
		"provider":    provider,
	}
	var out map[string]any
	if err := v.c.T.JSON(ctx, "cloud.VM.Status", "POST", url, body, &out); err != nil {
		return nil, err
	}
	return out, nil
}

// Action issues a lifecycle action against a previously-provisioned VM.
// Valid actions are "start", "stop", "restart", and "reboot" — anything
// else returns a ValidationError before any network round-trip.
//
// Maps to POST {NodeURL}/api/v2/provision/vm/action.
func (v *VM) Action(ctx context.Context, input ActionInput) (map[string]any, error) {
	if input.InstanceID == "" {
		return nil, &vxerrors.ValidationError{Failure: &vxerrors.Failure{
			Op: "cloud.VM.Action", Message: "InstanceID is required",
		}}
	}
	switch input.Action {
	case "start", "stop", "restart", "reboot":
		// ok
	default:
		return nil, &vxerrors.ValidationError{Failure: &vxerrors.Failure{
			Op: "cloud.VM.Action",
			Message: fmt.Sprintf(
				"Action must be one of start/stop/restart/reboot (got %q)",
				input.Action),
		}}
	}
	provider := input.Cloud
	if provider == "" {
		provider = "aws"
	}
	url := transport.JoinURL(v.c.NodeURL, "/api/v2/provision/vm/action")
	body := map[string]any{
		"instance_id": input.InstanceID,
		"action":      input.Action,
		"provider":    provider,
	}
	var out map[string]any
	if err := v.c.T.JSON(ctx, "cloud.VM.Action", "POST", url, body, &out); err != nil {
		return nil, err
	}
	return out, nil
}

// ── Network (VPC) ────────────────────────────────────────────────

type Network struct{ c *Client }

// CreateVPC provisions a VPC.
func (n *Network) CreateVPC(ctx context.Context, name, cloud, region string) (*Result, error) {
	return n.c.commonProvision(ctx,
		"cloud.Network.CreateVPC",
		"/api/v2/tenant/provision/networks",
		name, cloud, region, "", "vpc", nil)
}

// ── Database (RDS / Aurora / ElastiCache Redis) ───────────────────

type Database struct{ c *Client }

// DBOpts describes a managed database provision request.
type DBOpts struct {
	Engine                  string // "postgres" | "postgresql" | "mysql" | "mariadb"
	EngineVersion           string
	InstanceClass           string
	StorageGB               int
	Username                string
	Password                string
	DBName                  string
	Port                    int
	VPCID                   string
	SubnetIDs               []string
	AllowedSecurityGroupIDs []string
	VPCSecurityGroupIDs     []string
	InstanceCount           int
	NodeType                string
	NumCacheNodes           int
	PubliclyAccessible      bool
	MultiAZ                 bool
	BackupRetention         int
	Encryption              bool
}

// CreateRDS provisions an RDS instance.
func (d *Database) CreateRDS(ctx context.Context, name, region string, opts DBOpts) (*Result, error) {
	return d.c.commonProvision(ctx,
		"cloud.Database.CreateRDS",
		"/api/v2/tenant/provision/databases",
		name, "aws", region, "", "rds",
		map[string]interface{}{"configuration": databaseConfig(name, region, "rds", opts)},
	)
}

// CreateAurora provisions an Aurora cluster. Pass at least two SubnetIDs.
func (d *Database) CreateAurora(ctx context.Context, name, region string, opts DBOpts) (*Result, error) {
	return d.c.commonProvision(ctx,
		"cloud.Database.CreateAurora",
		"/api/v2/tenant/provision/databases",
		name, "aws", region, "", "aurora",
		map[string]interface{}{"configuration": databaseConfig(name, region, "aurora", opts)},
	)
}

// CreateRedis provisions an ElastiCache Redis cluster.
func (d *Database) CreateRedis(ctx context.Context, name, region string, opts DBOpts) (*Result, error) {
	return d.c.commonProvision(ctx,
		"cloud.Database.CreateRedis",
		"/api/v2/tenant/provision/databases",
		name, "aws", region, "", "redis",
		map[string]interface{}{"configuration": databaseConfig(name, region, "redis", opts)},
	)
}

func databaseConfig(name, region, resourceType string, opts DBOpts) map[string]interface{} {
	cfg := map[string]interface{}{
		"region": region,
	}
	switch resourceType {
	case "redis":
		nodeType := opts.NodeType
		if nodeType == "" {
			nodeType = "cache.t3.micro"
		}
		numCacheNodes := opts.NumCacheNodes
		if numCacheNodes == 0 {
			numCacheNodes = 1
		}
		cfg["node_type"] = nodeType
		cfg["num_cache_nodes"] = numCacheNodes
		if len(opts.SubnetIDs) > 0 {
			cfg["subnet_ids"] = opts.SubnetIDs
		}
		if len(opts.VPCSecurityGroupIDs) > 0 {
			cfg["vpc_security_group_ids"] = opts.VPCSecurityGroupIDs
		}
		return cfg
	default:
		engine := opts.Engine
		if engine == "" {
			engine = "mysql"
		}
		version := opts.EngineVersion
		if version == "" {
			version = "8.0"
		}
		instanceClass := opts.InstanceClass
		if instanceClass == "" {
			instanceClass = "db.t3.micro"
			if resourceType == "aurora" {
				instanceClass = "db.t3.medium"
			}
		}
		storageGB := opts.StorageGB
		if storageGB == 0 {
			storageGB = 20
		}
		port := opts.Port
		if port == 0 {
			port = 3306
			if engine == "postgres" || engine == "postgresql" {
				port = 5432
			}
		}
		dbName := opts.DBName
		if dbName == "" {
			dbName = name
		}
		username := opts.Username
		if username == "" {
			username = "admin"
		}
		cfg["engine"] = engine
		cfg["version"] = version
		cfg["instance_type"] = instanceClass
		cfg["storage_size"] = storageGB
		cfg["multi_az"] = opts.MultiAZ
		cfg["backup_retention"] = opts.BackupRetention
		if opts.BackupRetention == 0 {
			cfg["backup_retention"] = 7
		}
		cfg["encryption"] = true
		cfg["publicly_accessible"] = opts.PubliclyAccessible
		cfg["username"] = username
		cfg["password"] = opts.Password
		cfg["db_name"] = dbName
		cfg["port"] = port
		cfg["vpc_id"] = opts.VPCID
		if resourceType == "aurora" {
			cfg["instance_count"] = opts.InstanceCount
			if opts.InstanceCount == 0 {
				cfg["instance_count"] = 2
			}
		}
		if len(opts.SubnetIDs) > 0 {
			cfg["subnet_ids"] = opts.SubnetIDs
		}
		if len(opts.AllowedSecurityGroupIDs) > 0 {
			cfg["allowed_security_group_ids"] = opts.AllowedSecurityGroupIDs
		}
		return cfg
	}
}

// ── Kubernetes (EKS / GKE) ────────────────────────────────────────

type Kubernetes struct{ c *Client }

// CreateCluster provisions a managed Kubernetes cluster.
func (k *Kubernetes) CreateCluster(ctx context.Context, name, cloud, region string) (*Result, error) {
	return k.c.commonProvision(ctx,
		"cloud.Kubernetes.CreateCluster",
		"/api/v2/tenant/provision/kubernetes",
		name, cloud, region, "", "eks", nil)
}

// ── Serverless (Lambda / Cloud Functions) ────────────────────────

type Serverless struct{ c *Client }

// CreateFunction provisions a serverless function.
func (s *Serverless) CreateFunction(ctx context.Context, name, cloud, region, runtime string) (*Result, error) {
	return s.c.commonProvision(ctx,
		"cloud.Serverless.CreateFunction",
		"/api/v2/tenant/provision/serverless",
		name, cloud, region, "", "lambda",
		map[string]interface{}{"runtime": runtime},
	)
}

File: typescript/src/cloud.ts
/**
 * Cloud resource — VM / IAM / S3 / Database / Network / Kubernetes / Serverless.
 *
 * Thin in v0.1 — covers the same endpoints the Go and Python SDKs do.
 * Full coverage tracked in BIG_PLAN.md M4 (Cloud full coverage).
 */

import type { Transport } from './transport.js';

type CloudProvider = 'aws' | 'gcp' | 'azure';
type ManagedDatabaseType = 'rds' | 'aurora' | 'redis' | 'elasticache';

interface ManagedDatabaseProvisionInput {
  name?: string;
  resourceName?: string;
  cloud?: CloudProvider;
  resourceType?: ManagedDatabaseType;
  engine?: 'postgresql' | 'postgres' | 'mysql' | 'mariadb' | 'mongodb';
  instanceClass?: string;
  storageGb?: number;
  region: string;
  configuration?: Record<string, unknown>;
  tags?: Record<string, string>;
}

export class Cloud {
  readonly vm: VM;
  readonly s3: S3;
  readonly iam: IAM;
  readonly database: Database;
  readonly kubernetes: Kubernetes;
  readonly network: Network;
  readonly serverless: Serverless;

  constructor(private readonly t: Transport) {
    this.vm = new VM(t);
    this.s3 = new S3(t);
    this.iam = new IAM(t);
    this.database = new Database(t);
    this.kubernetes = new Kubernetes(t);
    this.network = new Network(t);
    this.serverless = new Serverless(t);
  }

  // Flat shortcut that mirrors the Python SDK's `c.cloud.create_vm(...)`.
  // Forwards to `this.vm.provision(...)`; identical surface, just lets users
  // who reach for the Python shape from TS land in the same place.
  async createVm(input: VMProvisionInput): Promise<Record<string, unknown>> {
    return this.vm.provision(input);
  }
}

export interface VMProvisionInput {
  name?: string;
  cloud?: 'aws' | 'gcp' | 'azure';
  instanceType: string;
  region: string;
  keyPairName?: string;
  volumeSize?: number;
  tags?: Record<string, string>;
  ami?: string;
}

export class VM {
  constructor(private t: Transport) {}

  async provision(input: VMProvisionInput): Promise<Record<string, unknown>> {
    const cloud = input.cloud ?? 'aws';
    // When the caller passes `name`, build the verbose body the Python SDK
    // uses (and that prod has been verified against). When not, keep the
    // slim body that existing TS callers already rely on — no behavior
    // change for code that doesn't opt in to the named shape.
    const body: Record<string, unknown> = input.name
      ? {
          app_name: input.name,
          resource_name: input.name,
          instance_name: input.name,
          network_name: input.name,
          key_name: input.name,
          role_name: input.name,
          hostname: input.name,
          cloud_provider: cloud,
          provider: cloud,
          region: input.region,
          environment: 'development',
          resource_type: 'vm',
          instance_type: input.instanceType,
          volume_size: input.volumeSize ?? 30,
          key_pair_name: input.keyPairName,
          tags: input.tags ?? {},
          ami: input.ami,
        }
      : {
          provider: cloud,
          instance_type: input.instanceType,
          region: input.region,
          key_pair_name: input.keyPairName,
          tags: input.tags ?? {},
          ami: input.ami,
        };
    const res = await this.t.postJSON<Record<string, unknown>>(
      '/api/v2/tenant/provision/vm',
      body,
    );
    return res.body ?? {};
  }

  async status(input: { instanceId: string; cloud?: 'aws' | 'gcp' | 'azure' }): Promise<Record<string, unknown>> {
    const res = await this.t.postJSON<Record<string, unknown>>(
      '/api/v2/provision/vm/status',
      { instance_id: input.instanceId, provider: input.cloud ?? 'aws' },
    );
    return res.body ?? {};
  }

  async action(input: {
    instanceId: string;
    action: 'start' | 'stop' | 'restart' | 'reboot';
    cloud?: 'aws' | 'gcp' | 'azure';
  }): Promise<Record<string, unknown>> {
    const res = await this.t.postJSON<Record<string, unknown>>(
      '/api/v2/provision/vm/action',
      { instance_id: input.instanceId, action: input.action, provider: input.cloud ?? 'aws' },
    );
    return res.body ?? {};
  }
}

export class S3 {
  constructor(private t: Transport) {}

  async createBucket(input: { name: string; region?: string; cloud?: 'aws' | 'gcp' | 'azure' }): Promise<Record<string, unknown>> {
    const res = await this.t.postJSON<Record<string, unknown>>(
      '/api/v2/tenant/provision/s3',
      { bucket_name: input.name, region: input.region ?? 'us-east-1', provider: input.cloud ?? 'aws' },
    );
    return res.body ?? {};
  }
}

export class IAM {
  constructor(private t: Transport) {}

  async createPolicy(input: { name: string; document: string | Record<string, unknown> }): Promise<Record<string, unknown>> {
    const res = await this.t.postJSON<Record<string, unknown>>(
      '/api/v2/tenant/provision/iam/policy',
      {
        name: input.name,
        policy_document: typeof input.document === 'string' ? input.document : JSON.stringify(input.document),
      },
    );
    return res.body ?? {};
  }

  async createRole(input: { name: string; assumeRolePolicy: string | Record<string, unknown> }): Promise<Record<string, unknown>> {
    const res = await this.t.postJSON<Record<string, unknown>>(
      '/api/v2/tenant/provision/iam/role',
      {
        name: input.name,
        assume_role_policy: typeof input.assumeRolePolicy === 'string'
          ? input.assumeRolePolicy
          : JSON.stringify(input.assumeRolePolicy),
      },
    );
    return res.body ?? {};
  }

  /** Provision a key pair. Private key is stored in the workspace vault
   *  under `name`; the public material is uploaded to the cloud. */
  async createKeypair(input: { name: string; region?: string; cloud?: CloudProvider }): Promise<Record<string, unknown>> {
    const res = await this.t.postJSON<Record<string, unknown>>(
      '/api/v2/tenant/provision/security',
      commonProvisionBody({
        name: input.name,
        cloud: input.cloud ?? 'aws',
        region: input.region ?? 'us-east-1',
        resourceType: 'keypair',
      }),
    );
    return res.body ?? {};
  }
}

export class Database {
  constructor(private t: Transport) {}

  async provision(input: ManagedDatabaseProvisionInput): Promise<Record<string, unknown>> {
    const resourceType = input.resourceType ?? 'rds';
    const name = input.name ?? input.resourceName ?? String(input.configuration?.db_name ?? `${resourceType}-database`);
    const configuration: Record<string, unknown> = {
      region: input.region,
      ...(input.engine ? { engine: input.engine } : {}),
      ...(input.instanceClass ? { instance_type: input.instanceClass } : {}),
      ...(input.storageGb ? { storage_size: input.storageGb } : {}),
      ...(input.configuration ?? {}),
    };
    const res = await this.t.postJSON<Record<string, unknown>>(
      '/api/v2/tenant/provision/databases',
      {
        cloud_provider: input.cloud ?? 'aws',
        resource_type: resourceType,
        resource_name: name,
        configuration,
        tags: input.tags ?? { Name: name, ManagedBy: '@vxcloud/sdk' },
      },
    );
    return res.body ?? {};
  }

  async createRDS(input: Omit<ManagedDatabaseProvisionInput, 'resourceType'>): Promise<Record<string, unknown>> {
    return this.provision({
      ...input,
      resourceType: 'rds',
      configuration: {
        engine: input.engine ?? 'mysql',
        version: input.configuration?.version ?? '8.0',
        instance_type: input.instanceClass ?? input.configuration?.instance_type ?? 'db.t3.micro',
        storage_size: input.storageGb ?? input.configuration?.storage_size ?? 20,
        username: input.configuration?.username ?? 'admin',
        db_name: input.configuration?.db_name ?? input.name ?? input.resourceName,
        port: input.configuration?.port ?? ((input.engine === 'postgres' || input.engine === 'postgresql') ? 5432 : 3306),
        ...input.configuration,
      },
    });
  }

  async createAurora(input: Omit<ManagedDatabaseProvisionInput, 'resourceType'>): Promise<Record<string, unknown>> {
    return this.provision({
      ...input,
      resourceType: 'aurora',
      configuration: {
        engine: input.engine ?? 'mysql',
        version: input.configuration?.version ?? '8.0',
        instance_type: input.instanceClass ?? input.configuration?.instance_type ?? 'db.t3.medium',
        instance_count: input.configuration?.instance_count ?? 2,
        username: input.configuration?.username ?? 'admin',
        db_name: input.configuration?.db_name ?? input.name ?? input.resourceName,
        port: input.configuration?.port ?? ((input.engine === 'postgres' || input.engine === 'postgresql') ? 5432 : 3306),
        ...input.configuration,
      },
    });
  }

  async createRedis(input: Omit<ManagedDatabaseProvisionInput, 'resourceType' | 'engine' | 'instanceClass' | 'storageGb'>): Promise<Record<string, unknown>> {
    return this.provision({
      ...input,
      resourceType: 'redis',
      configuration: {
        node_type: input.configuration?.node_type ?? 'cache.t3.micro',
        num_cache_nodes: input.configuration?.num_cache_nodes ?? 1,
        ...input.configuration,
      },
    });
  }
}

export class Kubernetes {
  constructor(private t: Transport) {}

  async provision(input: {
    cloud: 'aws' | 'gcp' | 'azure';
    clusterName: string;
    nodeCount: number;
    nodeType: string;
    region: string;
  }): Promise<Record<string, unknown>> {
    const res = await this.t.postJSON<Record<string, unknown>>(
      '/api/v2/tenant/provision/kubernetescluster/deploy',
      {
        provider: input.cloud,
        cluster_name: input.clusterName,
        node_count: input.nodeCount,
        node_type: input.nodeType,
        region: input.region,
      },
    );
    return res.body ?? {};
  }

  async listClusters(): Promise<unknown[]> {
    const res = await this.t.get<{ clusters?: unknown[] } | unknown[]>('/api/v2/tenant/kubernetes/clusters');
    if (Array.isArray(res.body)) return res.body;
    return (res.body as { clusters?: unknown[] })?.clusters ?? [];
  }

  async clusterDetails(name: string): Promise<Record<string, unknown>> {
    const res = await this.t.postJSON<Record<string, unknown>>(
      '/api/v2/tenant/kubernetes/cluster/details', { name },
    );
    return res.body ?? {};
  }
}

/** VPC / network provisioning. Mirrors Go `cloud.Network`. */
export class Network {
  constructor(private t: Transport) {}

  async createVpc(input: { name: string; cloud?: CloudProvider; region?: string }): Promise<Record<string, unknown>> {
    const res = await this.t.postJSON<Record<string, unknown>>(
      '/api/v2/tenant/provision/networks',
      commonProvisionBody({
        name: input.name,
        cloud: input.cloud ?? 'aws',
        region: input.region ?? 'us-east-1',
        resourceType: 'vpc',
      }),
    );
    return res.body ?? {};
  }
}

/** Lambda / Cloud Functions provisioning. Mirrors Go `cloud.Serverless`. */
export class Serverless {
  constructor(private t: Transport) {}

  async createFunction(input: {
    name: string;
    cloud?: CloudProvider;
    region?: string;
    runtime: string;
  }): Promise<Record<string, unknown>> {
    if (!input.runtime) throw new Error('serverless.createFunction: runtime is required');
    const res = await this.t.postJSON<Record<string, unknown>>(
      '/api/v2/tenant/provision/serverless',
      {
        ...commonProvisionBody({
          name: input.name,
          cloud: input.cloud ?? 'aws',
          region: input.region ?? 'us-east-1',
          resourceType: 'lambda',
        }),
        runtime: input.runtime,
      },
    );
    return res.body ?? {};
  }
}

/** Build the canonical envelope `/api/v2/tenant/provision/*` handlers expect.
 *  Mirrors Go `commonProvision` and Python `Cloud._provision`. */
function commonProvisionBody(input: {
  name: string;
  cloud: CloudProvider;
  region: string;
  resourceType: string;
  env?: string;
}): Record<string, unknown> {
  return {
    app_name: input.name,
    resource_name: input.name,
    instance_name: input.name,
    network_name: input.name,
    key_name: input.name,
    role_name: input.name,
    hostname: input.name,
    cloud_provider: input.cloud,
    region: input.region,
    environment: input.env ?? 'development',
    resource_type: input.resourceType,
  };
}
